### PR TITLE
Improved CI package manager detection

### LINKED ISSUE #1312: Recent Buildpack change incorrectly uses pnpm despite Yarn configuration
**Describe the bug**

The Heroku Node.js buildpack is failing CI test run when using Yarn as the package manager, despite the project being configured correctly for it. The build output shows a "Usage Error: This project is configured to use yarn" message, followed by an attempt to run tests with `pnpm`.

The expected behavior is for the buildpack to detect Yarn as the package manager and execute tests using Yarn.

Log output:

```
-----> Running Node.js buildpack tests...
Usage Error: This project is configured to use yarn
$ pnpm ...
-----> Node.js buildpack tests failed with exit status 1
```

**To Reproduce**

1. Use Heroku Node.js buildpack on a project with Yarn as the package manager.
2. Have `yarn` declared as the package manager in package.json using `"packageManager": "yarn@1.22.22"`
4. Trigger a CI test run on Heroku.
5. Observe the test run failing with the "Usage Error" message.

**Versions**

* Heroku Stack: Heroku-22
* Node Version: 20.10.0
* Yarn Version: 1.22.22
* Buildpack Version: heroku/nodejs (latest) 263

**Additional context**

* The issue started occurring after [a recent change in the Heroku buildpack that added support for `pnpm`.](https://github.com/heroku/heroku-buildpack-nodejs/commit/976d5e0adbf08d2ab6425cc876aebae6983a4850)
* I was able to fix the issue using the following diff, but this shouldn't be necessary.

```diff
diff --git a/package.json b/package.json
index 77e1751..b6a362c 100644
--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
     "nightly": "yarn workspace backend run nightly"
   },
   "engines": {
-    "node": "20.10.x"
-  },
-  "packageManager": "yarn@1.22.22"
+    "node": "20.10.x",
+    "yarn": "1.22.22"
+  }
 }
```

---

**Potential Root Cause and Solution**

It seems the buildpack is prioritizing `pnpm` even when the project is explicitly configured to use Yarn. The fix of adding `"yarn": "1.22.22"` to the `engines` section reinforces this theory.

The buildpack likely needs to improve its package manager detection logic to ensure it respects the project's configuration and doesn't default to `pnpm` when Yarn is specified. 

Please let me know if you need any more help refining this issue or if there's any other information I can provide!

### PR DESCRIPTION
**Pull request by colincasey**
The previous method for detecting package managers works well when `yarn`, `pnpm`, or `npm` binaries are installed in the traditional way but fails when `corepack` shims are present. This PR changes the detection logic to use package metadata when determining which package manager to use.

Fixes #1312

### CODE REVIEW

File: bin/test [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+
+which_tool() {
+  yarn_engine=$(read_package_json ".engines.yarn")
+  pnpm_engine=$(read_package_json ".engines.pnpm")
+  package_manager=$(read_package_json ".packageManager")
+  if [ -n "$pnpm_engine" ] || [[ "$package_manager" == pnpm* ]]; then
+    echo "pnpm"
+  elif [ -n "$yarn_engine" ] || [[ "$package_manager" == yarn* ]]; then
+    echo "yarn"
+  else
```
Comment (joshwlewis) [inline bin/test:19]: Should we handle conflicts in some way? What if both `engines.yarn` and `engines.pnpm` are set? What if `engines.yarn` is set, but `packageManager` is `pnpm@*`?

**Review by joshwlewis — APPROVED** [review]
Looks like this will fix the issue, so leaving a 👍🏻. I think we probably could be more robust here by detecting conflicts, but I think we've already got that handled in `bin/compile` which will absolutely run before `bin/test` anyway.

File: bin/test (hunk 1 above)
Comment (colincasey) [reply bin/test:19]: Good question. I was thinking that case would result in an error earlier in the `compile` (or `test-compile`) process but we only check and fail if there are multiple lockfiles, not conflicting package manager metadata. I'll add some coverage for this possibility too.

Comment (colincasey) [reply bin/test:19]: The above should be addressed by #1317

**Pull request merged**